### ISSUE TITLE
`Process.Kill(entireProcessTree: true)` does not fire `Exited` event on Windows

### ISSUE THREAD

**Issue by alexrp** [dotnet/runtime#63328]
### Description

The `Kill(bool entireProcessTree)` overload on `System.Diagnostics.Process` does not cause the `Exited` event to be fired on Windows when `entireProcessTree` is `true`, even when `EnableRaisingEvents` is `true` and `WaitForExit()` is called.

### Reproduction Steps

```csharp
using System.Diagnostics;

var proc = new Process
{
    StartInfo = new()
    {
        FileName = "bash",
    },
    EnableRaisingEvents = true,
};

proc.Exited += (_, _) => Console.WriteLine("\nExited");
proc.Start();

Thread.Sleep(5000);

proc.Kill(true);
proc.WaitForExit();
```

Run the program and wait for the Bash child process to be killed.

### Expected behavior

The program should print `Exited`.

### Actual behavior

Nothing is printed; the `Exited` event is not fired.

### Known Workarounds

`Kill(false)` will fire `Exited` as expected, but is obviously not functionally equivalent.

### Configuration

The issue happens on Windows; on Linux, `Exited` is raised as expected. Not sure about macOS.

```
$ dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.101
 Commit:    ef49f6213a

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22523
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\program files\dotnet\sdk\6.0.101\

Host (useful for support):
  Version: 6.0.1
  Commit:  3a25a7f1cc
```

**Comment by danmoseley** [comment]
Thanks for the report @alexrp . Any interest in deciding a little and perhaps a PR?

**Comment by stephentoub** [comment]
@alexrp, FWIW, I tried your repro and it successfully prints out Exited for me.

**Comment by KalleOlaviNiemitalo** [comment]
Perhaps a race condition between the event handler running and the main thread exiting.

**Comment by alexrp** [comment]
Huh, I thought I had managed to isolate a valid repro but Stephen is right. Running it under CMD or Git Bash (mintty), the `Exited` message is printed. I was running it in Git Bash under Windows Terminal, and for whatever obscure reason, the string has to be modified from `"\nExited"` to `"\nExited\n"` for it to show up correctly in that environment for me. (This is probably unrelated to .NET.)

Having said that, the actual issue I described still exists and causes breakage as described in https://github.com/alexrp/system-terminal/issues/72 but it seems there's more to triggering the issue than I thought. Will investigate again and post an update.

**Comment by alexrp** [comment]
No repro yet, but I did figure out that I was accidentally swallowing an interesting exception:

```
System.InvalidOperationException: No process is associated with this object.
   at System.Diagnostics.Process.EnsureState(State state)
   at System.Diagnostics.Process.EnsureState(State state)
   at System.Diagnostics.Process.GetProcessHandle(Int32 access, Boolean throwIfExited)
   at System.Diagnostics.Process.WaitForExitCore(Int32 milliseconds)
   at System.Diagnostics.Process.WaitForExit()
   at System.Processes.ChildProcess.Kill(Boolean entireProcessTree) in C:\Users\alex\source\repos\system-terminal\src\core\Processes\ChildProcess.cs:line 185
   at System.Processes.ChildProcess.<>c.<.ctor>b__17_2(Object state, CancellationToken token) in C:\Users\alex\source\repos\system-terminal\src\core\Processes\ChildProcess.cs:line 129
```

Something certainly seems wrong if `Kill(true)` leaves the `Process` object unassociated; [`Kill` documentation](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.kill?view=net-6.0) explicitly states that calling `WaitForExit` afterwards is a supported scenario.

(Side note: [`WaitForExit` documentation](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.waitforexit?view=net-6.0#system-diagnostics-process-waitforexit) could probably do with some improvement... documenting that it can throw the `SystemException` base type is not super helpful.)

**Comment by alexrp** [comment]
Good news: There is no .NET bug after all.

It turns out I had a race condition where I ended up disposing the `Process` from thread A while thread B was calling `KIll` + `WaitForExit`. This would lead to the above `InvalidOperationException` (which could be thrown from either `Kill` or `WaitForExit`). I would have expected to see `ObjectDisposedException` in a case like this, so it took a bit for the realization to set in that it might have something to do with disposal.

I've fixed the race condition in my code and all is well. Sorry for the noise. (I do wonder if the `InvalidOperationException` (as opposed to `ObjectDisposedException`) is expected behavior here, but I suppose that would be a separate issue.)
